apiserver/facades/agent/uniter/uniter.go

```diff
@@ -875,35 +874,40 @@ func (u *UniterAPI) CharmArchiveSha256(ctx context.Context, args params.CharmURL
 }
 
 func (u *UniterAPI) oneCharmArchiveSha256(ctx context.Context, curl string) (string, error) {
-	// The charm in state may only be a placeholder when this call is made.
-	// Ideally, the unit agent would not be started until the charm is fully available,
-	// but that's not currently the case and it doesn't hurt to be defensive here regardless.
-	// We'll retry the sha256 lookup if the charm is still pending and therefore not found.
-	var sha string
-	err := retry.Call(retry.CallArgs{
-		Func: func() error {
-			sch, err := u.st.Charm(curl)
-			if err != nil {
-				return errors.Trace(err)
-			}
-			sha = sch.BundleSha256()
-			if sha == "" {
-				return errors.NotFoundf("downloaded charm %q", curl)
-			}
-			return nil
-		},
-		IsFatalError: func(err error) bool {
-			return !errors.Is(err, errors.NotFound)
-		},
-		Stop:     ctx.Done(),
-		Delay:    3 * time.Second,
-		Attempts: 20,
-		Clock:    u.clock,
+	cu, err := charm.ParseURL(curl)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	source := domaincharm.LocalSource
+	if charm.CharmHub.Matches(cu.Schema) {
+		source = domaincharm.CharmHubSource
+	}
```

Comment:
`ParseCharmSchema` from `domain/application/charm/types.go` would probably be better here. Maybe defaulting to `LocalSource` in the `""` case (if it's needed?)

Reply:
Ah, good shout!